File: desktop_app/src/services/deviceManager.js
const EventEmitter = require('events');
const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const crypto = require('crypto');
const CertificateManager = require('./certificateManager');

class DeviceManager extends EventEmitter {
  constructor() {
    super();
    this.devices = new Map();
    this.localDeviceId = this.generateDeviceId();
    this.localDeviceName = os.hostname();
    this.trustedDevices = new Map();
    this.deviceAliases = new Map();
    this.certificateManager = new CertificateManager();
    this.settings = {
      streamCount: 'auto',
      language: 'en',
      notifications: true,
      notificationSound: true,
      autoAcceptTrusted: true,
      theme: 'auto',
      downloadPath: path.join(os.homedir(), 'Downloads')
    };
    this.configPath = path.join(os.homedir(), '.rapidtransfer');
    this.init();
  }

  async init() {
    try {
      await fs.mkdir(this.configPath, { recursive: true });
      await this.certificateManager.initialize();
      await this.loadConfig();
    } catch (err) {
      console.error('Failed to initialize device manager:', err);
    }
  }

  generateDeviceId() {
    return crypto.randomBytes(16).toString('hex');
  }

  async loadConfig() {
    try {
      const configFile = path.join(this.configPath, 'config.json');
      const data = await fs.readFile(configFile, 'utf-8');
      const config = JSON.parse(data);
      
      this.localDeviceId = config.deviceId || this.localDeviceId;
      this.localDeviceName = config.deviceName || this.localDeviceName;
      this.trustedDevices = new Map(Object.entries(config.trustedDevices || {}));
      this.deviceAliases = new Map(Object.entries(config.deviceAliases || {}));
      this.settings = { ...this.settings, ...config.settings };
    } catch (err) {
      if (err.code !== 'ENOENT') {
        console.error('Failed to load config:', err);
      }
      await this.saveConfig();
    }
  }

  async saveConfig() {
    try {
      const configFile = path.join(this.configPath, 'config.json');
      const config = {
        deviceId: this.localDeviceId,
        deviceName: this.localDeviceName,
        trustedDevices: Object.fromEntries(this.trustedDevices),
        deviceAliases: Object.fromEntries(this.deviceAliases),
        settings: this.settings
      };
      await fs.writeFile(configFile, JSON.stringify(config, null, 2));
    } catch (err) {
      console.error('Failed to save config:', err);
    }
  }

  getLocalDeviceId() {
    return this.localDeviceId;
  }

  getLocalDeviceName() {
    return this.localDeviceName;
  }

  async setLocalDeviceName(name) {
    this.localDeviceName = name;
    await this.saveConfig();
    return true;
  }

  addDevice(device) {
    this.devices.set(device.id, {
      ...device,
      lastSeen: Date.now(),
      alias: this.deviceAliases.get(device.id) || null,
      trusted: this.trustedDevices.has(device.id)
    });
    this.emit('device-found', this.devices.get(device.id));
  }

  removeDevice(deviceId) {
    if (this.devices.has(deviceId)) {
      this.devices.delete(deviceId);
      this.emit('device-lost', deviceId);
    }
  }

  getDevice(deviceId) {
    return this.devices.get(deviceId);
  }

  getDevices() {
    return Array.from(this.devices.values());
  }

  async trustDevice(deviceId, publicKey) {
    this.trustedDevices.set(deviceId, {
      publicKey,
      trustedAt: Date.now()
    });
    const device = this.devices.get(deviceId);
    if (device) {
      device.trusted = true;
    }
    await this.saveConfig();
  }

  async forgetDevice(deviceId) {
    this.trustedDevices.delete(deviceId);
    this.deviceAliases.delete(deviceId);
    const device = this.devices.get(deviceId);
    if (device) {
      device.trusted = false;
      device.alias = null;
    }
    await this.saveConfig();
    return true;
  }

  async setDeviceAlias(deviceId, alias) {
    if (alias) {
      this.deviceAliases.set(deviceId, alias);
    } else {
      this.deviceAliases.delete(deviceId);
    }
    const device = this.devices.get(deviceId);
    if (device) {
      device.alias = alias;
    }
    await this.saveConfig();
    return true;
  }

  isTrusted(deviceId) {
    return this.trustedDevices.has(deviceId);
  }

  getTrustedDeviceKey(deviceId) {
    const trusted = this.trustedDevices.get(deviceId);
    return trusted ? trusted.publicKey : null;
  }

  getSettings() {
    return { ...this.settings };
  }

  async updateSettings(newSettings) {
    this.settings = { ...this.settings, ...newSettings };
    await this.saveConfig();
    return true;
  }

  // Clean up old devices that haven't been seen in a while
  cleanupOldDevices(timeoutMs = 300000) { // 5 minutes default
    const now = Date.now();
    for (const [deviceId, device] of this.devices.entries()) {
      if (now - device.lastSeen > timeoutMs) {
        this.removeDevice(deviceId);
      }
    }
  }
}

module.exports = DeviceManager;

File: mobile_app/lib/services/device_manager.dart
/**
 * Rapid Transfer - Device Manager
 * Copyright (C) 2025 Seth Johnston - Licensed under AGPL-3.0
 */

import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../models/device.dart';

class DeviceManager extends ChangeNotifier {
  final Map<String, Device> _devices = {};
  String? _localDeviceId;
  String _localDeviceName = '';
  Map<String, String> _deviceAliases = {};
  Set<String> _trustedDevices = {};

  // Settings
  String _streamCount = 'auto';
  String _language = 'en';
  bool _notifications = true;
  bool _keepAwake = true;
  bool _autoConfigureHotspot = true;

  List<Device> get devices => _devices.values.toList();
  String get localDeviceName => _localDeviceName;
  String? getLocalDeviceId() => _localDeviceId;
  String get streamCount => _streamCount;
  String get language => _language;
  bool get notifications => _notifications;
  bool get keepAwake => _keepAwake;
  bool get autoConfigureHotspot => _autoConfigureHotspot;

  Future<void> init() async {
    await _loadConfig();
  }

  Future<void> _loadConfig() async {
    final prefs = await SharedPreferences.getInstance();
    _localDeviceId = prefs.getString('deviceId');
    if (_localDeviceId == null) {
      _localDeviceId = _generateDeviceId();
      await prefs.setString('deviceId', _localDeviceId!);
    }

    _localDeviceName = prefs.getString('deviceName') ?? 'Android Device';
    _streamCount = prefs.getString('streamCount') ?? 'auto';
    _language = prefs.getString('language') ?? 'en';
    _notifications = prefs.getBool('notifications') ?? true;
    _keepAwake = prefs.getBool('keepAwake') ?? true;
    _autoConfigureHotspot = prefs.getBool('autoConfigureHotspot') ?? true;

    // Load trusted devices
    final trustedList = prefs.getStringList('trustedDevices') ?? [];
    _trustedDevices = trustedList.toSet();

    // Load aliases
    final aliasKeys = prefs.getKeys().where((k) => k.startsWith('alias_'));
    for (final key in aliasKeys) {
      final deviceId = key.substring(6);
      final alias = prefs.getString(key);
      if (alias != null) {
        _deviceAliases[deviceId] = alias;
      }
    }
  }

  String _generateDeviceId() {
    return DateTime.now().millisecondsSinceEpoch.toString();
  }

  void addDevice(Device device) {
    _devices[device.id] = device.copyWith(
      alias: _deviceAliases[device.id],
      trusted: _trustedDevices.contains(device.id),
    );
    notifyListeners();
  }

  void removeDevice(String deviceId) {
    _devices.remove(deviceId);
    notifyListeners();
  }

  Device? getDevice(String deviceId) {
    return _devices[deviceId];
  }

  Future<void> setDeviceName(String name) async {
    _localDeviceName = name;
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('deviceName', name);
    notifyListeners();
  }

  Future<void> trustDevice(String deviceId) async {
    _trustedDevices.add(deviceId);
    final device = _devices[deviceId];
    if (device != null) {
      _devices[deviceId] = device.copyWith(trusted: true);
    }
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList('trustedDevices', _trustedDevices.toList());
    notifyListeners();
  }

  Future<void> forgetDevice(String deviceId) async {
    _trustedDevices.remove(deviceId);
    _deviceAliases.remove(deviceId);
    final device = _devices[deviceId];
    if (device != null) {
      _devices[deviceId] = device.copyWith(trusted: false, alias: null);
    }
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList('trustedDevices', _trustedDevices.toList());
    await prefs.remove('alias_$deviceId');
    notifyListeners();
  }

  Future<void> setDeviceAlias(String deviceId, String? alias) async {
    final prefs = await SharedPreferences.getInstance();
    if (alias != null && alias.isNotEmpty) {
      _deviceAliases[deviceId] = alias;
      await prefs.setString('alias_$deviceId', alias);
    } else {
      _deviceAliases.remove(deviceId);
      await prefs.remove('alias_$deviceId');
    }
    final device = _devices[deviceId];
    if (device != null) {
      _devices[deviceId] = device.copyWith(alias: alias);
    }
    notifyListeners();
  }

  Future<void> updateSettings({
    String? streamCount,
    String? language,
    bool? notifications,
    bool? keepAwake,
    bool? autoConfigureHotspot,
  }) async {
    final prefs = await SharedPreferences.getInstance();

    if (streamCount != null) {
      _streamCount = streamCount;
      await prefs.setString('streamCount', streamCount);
    }
    if (language != null) {
      _language = language;
      await prefs.setString('language', language);
    }
    if (notifications != null) {
      _notifications = notifications;
      await prefs.setBool('notifications', notifications);
    }
    if (keepAwake != null) {
      _keepAwake = keepAwake;
      await prefs.setBool('keepAwake', keepAwake);
    }
    if (autoConfigureHotspot != null) {
      _autoConfigureHotspot = autoConfigureHotspot;
      await prefs.setBool('autoConfigureHotspot', autoConfigureHotspot);
    }

    notifyListeners();
  }

  void dispose() {
    super.dispose();
  }
}
